// @digdir/designsystemet-web/dist/cjs/breadcrumbs.cjs
const require_utils = require('./utils.cjs');

//#region src/breadcrumbs.ts
const ATTR_LABEL = "aria-label";
const ATTR_LABEL_HIDDEN = "data-label";
var DSBreadcrumbsElement = class extends require_utils.DSElement {
	_items;
	_unresize;
	_unmutate;
	static get observedAttributes() {
		return [ATTR_LABEL];
	}
	connectedCallback() {
		if (!require_utils.attr(this, ATTR_LABEL_HIDDEN)) require_utils.attrRequiredWarning(this, ATTR_LABEL);
		const render = require_utils.debounce(this.attributeChangedCallback.bind(this), 100);
		this._items = this.getElementsByTagName("a");
		this._unresize = require_utils.on(window, "resize", render);
		this._unmutate = require_utils.onMutation(this, render, {
			debounce: 0,
			childList: true,
			subtree: true
		});
	}
	attributeChangedCallback() {
		const last = this._items?.[this._items.length - 1];
		const lastInList = last?.parentElement === this ? null : last;
		const isListHidden = !lastInList?.offsetHeight;
		const labelHidden = require_utils.attr(this, ATTR_LABEL_HIDDEN);
		const label = require_utils.attr(this, ATTR_LABEL);
		require_utils.attr(this, "role", isListHidden ? null : "navigation");
		if (isListHidden && !labelHidden && label) {
			require_utils.attr(this, ATTR_LABEL_HIDDEN, label);
			require_utils.attr(this, ATTR_LABEL, null);
		} else if (!isListHidden && labelHidden && !label) {
			require_utils.attr(this, ATTR_LABEL, labelHidden);
			require_utils.attr(this, ATTR_LABEL_HIDDEN, null);
		}
		for (const item of this._items || []) require_utils.attr(item, "aria-current", item === lastInList ? "page" : null);
	}
	disconnectedCallback() {
		this._unresize?.();
		this._unmutate?.();
		this._unresize = this._unmutate = this._items = void 0;
	}
};
require_utils.customElements.define("ds-breadcrumbs", DSBreadcrumbsElement);

//#endregion
exports.DSBreadcrumbsElement = DSBreadcrumbsElement;
