// @digdir/designsystemet-web/dist/cjs/index.cjs
const require_utils = require('./utils.cjs');
const require_breadcrumbs = require('./breadcrumbs.cjs');
require('./clickdelegatefor.cjs');
require('./details.cjs');
require('./dialog.cjs');
const require_error_summary = require('./error-summary.cjs');
const require_field = require('./field.cjs');
const require_pagination = require('./pagination.cjs');
const require_suggestion = require('./suggestion.cjs');
const require_tabs = require('./tabs.cjs');
require('./popover.cjs');
require('./toggle-group.cjs');
require('./tooltip.cjs');

//#region src/index.ts
if (require_utils.isBrowser()) import("invokers-polyfill");

//#endregion
exports.DSBreadcrumbsElement = require_breadcrumbs.DSBreadcrumbsElement;
exports.DSErrorSummaryElement = require_error_summary.DSErrorSummaryElement;
exports.DSFieldElement = require_field.DSFieldElement;
exports.DSPaginationElement = require_pagination.DSPaginationElement;
exports.DSSuggestionElement = require_suggestion.DSSuggestionElement;
exports.DSTabElement = require_tabs.DSTabElement;
exports.DSTabListElement = require_tabs.DSTabListElement;
exports.DSTabPanelElement = require_tabs.DSTabPanelElement;
exports.DSTabsElement = require_tabs.DSTabsElement;
exports.pagination = require_pagination.pagination;
var _u_elements_u_datalist = require("@u-elements/u-datalist");
Object.keys(_u_elements_u_datalist).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return _u_elements_u_datalist[k]; }
  });
});
