// @digdir/designsystemet-web/dist/cjs/pagination.cjs
const require_utils = require('./utils.cjs');

//#region src/pagination.ts
const ATTR_LABEL = "aria-label";
const ATTR_CURRENT = "data-current";
const ATTR_TOTAL = "data-total";
const ATTR_HREF = "data-href";
const pagination = ({ current = 1, total = 10, show = 7 }) => ({
	prev: current > 1 ? current - 1 : 0,
	next: current < total ? current + 1 : 0,
	pages: getSteps(current, total, show).map((page, index) => ({
		current: page === current && "page",
		key: `key-${page}-${index}`,
		page
	}))
});
var DSPaginationElement = class extends require_utils.DSElement {
	_unmutate;
	static get observedAttributes() {
		return [ATTR_CURRENT, ATTR_TOTAL];
	}
	connectedCallback() {
		require_utils.attrRequiredWarning(this, ATTR_LABEL);
		if (require_utils.attr(this, ATTR_TOTAL)) require_utils.attrRequiredWarning(this, ATTR_CURRENT);
		if (require_utils.attr(this, ATTR_CURRENT)) require_utils.attrRequiredWarning(this, ATTR_TOTAL);
		this._unmutate = require_utils.onMutation(this, this.render.bind(this), {
			childList: true,
			subtree: true,
			debounce: 0
		});
	}
	disconnectedCallback() {
		this._unmutate?.();
		this._unmutate = void 0;
	}
	render() {
		const items = this.querySelectorAll("button,a");
		const href = require_utils.attr(this, ATTR_HREF);
		const CURRENT = require_utils.attr(this, ATTR_CURRENT) || void 0;
		const TOTAL = require_utils.attr(this, ATTR_TOTAL) || void 0;
		if (!CURRENT || !TOTAL) return;
		const { next, prev, pages } = pagination({
			current: parseInt(CURRENT, 10),
			total: parseInt(TOTAL, 10),
			show: items.length - 2
		});
		items.forEach((item, i) => {
			const page = i ? items[i + 1] ? pages[i - 1]?.page : next : prev;
			require_utils.attr(item, "aria-current", pages[i - 1]?.current ? "true" : null);
			require_utils.attr(item, "aria-hidden", page ? null : "true");
			require_utils.attr(item, "data-page", `${page}`);
			require_utils.attr(item, "tabindex", page ? null : "-1");
			if (item instanceof HTMLButtonElement) require_utils.attr(item, "value", `${page}`);
			if (href) require_utils.attr(item, "href", href.replace("$page", `${page}`));
		});
	}
};
require_utils.customElements.define("ds-pagination", DSPaginationElement);
function getSteps(now, max, show = Number.POSITIVE_INFINITY) {
	const offset = (show - 1) / 2;
	const start = Math.max(Math.min(now - Math.floor(offset), max - show + 1), 1);
	const end = Math.min(Math.max(now + Math.ceil(offset), show), max);
	const pages = Array.from({ length: end + 1 - start }, (_, i) => i + start);
	if (show > 4 && start > 1) pages.splice(0, 2, 1, 0);
	if (show > 3 && end < max) pages.splice(-2, 2, 0, max);
	return pages;
}

//#endregion
exports.DSPaginationElement = DSPaginationElement;
exports.pagination = pagination;
