// @digdir/designsystemet-web/dist/esm/breadcrumbs.js
import { DSElement, attr, attrRequiredWarning, customElements, debounce, on, onMutation } from "./utils.js";

//#region src/breadcrumbs.ts
const ATTR_LABEL = "aria-label";
const ATTR_LABEL_HIDDEN = "data-label";
var DSBreadcrumbsElement = class extends DSElement {
	_items;
	_unresize;
	_unmutate;
	static get observedAttributes() {
		return [ATTR_LABEL];
	}
	connectedCallback() {
		if (!attr(this, ATTR_LABEL_HIDDEN)) attrRequiredWarning(this, ATTR_LABEL);
		const render = debounce(this.attributeChangedCallback.bind(this), 100);
		this._items = this.getElementsByTagName("a");
		this._unresize = on(window, "resize", render);
		this._unmutate = onMutation(this, render, {
			debounce: 0,
			childList: true,
			subtree: true
		});
	}
	attributeChangedCallback() {
		const last = this._items?.[this._items.length - 1];
		const lastInList = last?.parentElement === this ? null : last;
		const isListHidden = !lastInList?.offsetHeight;
		const labelHidden = attr(this, ATTR_LABEL_HIDDEN);
		const label = attr(this, ATTR_LABEL);
		attr(this, "role", isListHidden ? null : "navigation");
		if (isListHidden && !labelHidden && label) {
			attr(this, ATTR_LABEL_HIDDEN, label);
			attr(this, ATTR_LABEL, null);
		} else if (!isListHidden && labelHidden && !label) {
			attr(this, ATTR_LABEL, labelHidden);
			attr(this, ATTR_LABEL_HIDDEN, null);
		}
		for (const item of this._items || []) attr(item, "aria-current", item === lastInList ? "page" : null);
	}
	disconnectedCallback() {
		this._unresize?.();
		this._unmutate?.();
		this._unresize = this._unmutate = this._items = void 0;
	}
};
customElements.define("ds-breadcrumbs", DSBreadcrumbsElement);

//#endregion
export { DSBreadcrumbsElement };
